### PR TITLE
WIP: Sampling, Spacing, Slicing, and BulkJobs

### PR DESCRIPTION
**Pull request by apoms**
This PR refactors Scanner to support:

- Sampling
- Spacing
- Slicing
- BulkJobs

NOTE: I'll finish up this description soon.

### CODE REVIEW

File: examples/shot_detection/shot_detect.py [hunk 1]
```diff
@@ -77,85 +77,128 @@ def place_image(i, fr):
...
         print('Computing a color histogram for each frame...')
-        frame = movie_table.as_op().all()
+        frame = db.ops.FrameInput()
         histogram = db.ops.Histogram(
             frame = frame,
             device = device)
-        job = Job(columns = [histogram], name = movie_name + '_hist')
-        hists_table = db.run(job, force=True)
-        print('\nTime: {:.1f}s'.format(time.time() - s))
+        output = db.ops.Output(columns=[histogram])
```
Comment (willcrichton) [inline examples/shot_detection/shot_detect.py]: Is there a good reason to have an explicit output node vs. just having `BulkJob(columns=[histogram], ...)`?

File: examples/tutorial/00_basic.py [hunk 2]
```diff
@@ -14,11 +14,12 @@
 
 # Initialize a connection to the Scanner database. Loads configuration from the
 # ~/.scanner.toml configuration file.
-with Database() as db:
+with Database(master="crissy:5001", workers=["crissy:5002"], debug=True) as db:
```
Comment (willcrichton) [inline examples/tutorial/00_basic.py]: Probably shouldn't have this hardcoded in the tutorial.

File: examples/tutorial/01_sampling.py [hunk 3]
```diff
@@ -1,29 +1,51 @@
...
+        num_rows += 1
+    assert num_rows == db.table('example').num_rows() / 8
 
     # Here's some examples of other sampling modes.
-    # Range takes a specific subset of a video. Here, it runs over all frames from
-    # 0 to 100
-    input_table.as_op().range(0, 100)
+    # Range takes a specific subset of a video. Here, it runs over all frames
+    # from 0 to 100
+    db.sampler.range(0, 100)
```
Comment (willcrichton) [inline examples/tutorial/01_sampling.py]: General docs question--should we have some kind of API reference, or just have these sorts of functions be documented in small examples via tutorials/applications?

File: examples/tutorial/03_ops.py [hunk 4]
```diff
@@ -21,8 +21,65 @@
...
+    # batch processing. If there are not enough elements left in a stream,
+    # the Op may receive less than a batch worth of elements.
+    histogram = db.ops.Histogram(
+        frame = frame,
+        batch = 8)
+
+
+    # Stencil
+    # Under construction...
+    diff = db.ops.FrameDifference(
```
Comment (willcrichton) [inline examples/tutorial/03_ops.py]: What's under construction?

File: python/scannerpy/collection.py [hunk 5]
```diff
@@ -28,4 +27,5 @@ def profiler(self):
         return self._db.profiler(self._descriptor.job_id)
 
     def as_op(self):
-        return SamplerOp(self)
+        #return SamplerOp(self)
```
Comment (willcrichton) [inline python/scannerpy/collection.py]: Are collections still functional or broken atm?

File: python/scannerpy/database.py [hunk 6]
```diff
@@ -1009,51 +970,76 @@ def eval_job(job_fn):
...
-            tasks = [task]
-            collection = input_op._collection
-            if collection is not None:
-                output_collection = job.name()
-                if self.has_collection(output_collection) and not force:
+        sorted_ops, input_ops, sampling_slicing_ops, output_ops = (
+            self._toposort(bulk_job.output()))
+
+        job_params = self.protobufs.BulkJobParameters()
+        job_name = ''.join(choice(ascii_uppercase) for _ in range(12))
```
Comment (willcrichton) [inline python/scannerpy/database.py]: Do jobs still need to have job names?

File: scanner/video/#video_decoder.cpp# [hunk 7]
```diff
@@ -0,0 +1,106 @@
+/* Copyright 2016 Carnegie Mellon University
```
Comment (willcrichton) [inline scanner/video/#video_decoder.cpp#]: Extra file?

File: tests/py_test.py [hunk 8]
```diff
@@ -119,16 +119,16 @@ def test_table_properties(db):
...
 
-def test_collection(db):
-    c = db.new_collection('test', ['test1', 'test2'])
-    frame = c.as_op().strided(2)
-    job = Job(
-        columns = [db.ops.Histogram(frame = frame)],
-        name = '_ignore')
-    db.run(job, show_progress=False, force=True)
-    db.delete_collection('test')
+# def test_collection(db):
```
Comment (willcrichton) [inline tests/py_test.py]: We should put a TODO to either revive this or remove collections entirely.

File: tests/py_test.py [hunk 9]
```diff
@@ -144,18 +144,130 @@ def test_load_video_column(db):
...
+    bulk_job = BulkJob(output=output_op, jobs=[job])
+    tables = db.run(bulk_job, force=True, show_progress=False)
+
+    num_rows = 0
+    for (frame_index, _) in tables[0].column('frame').load():
+        num_rows += 1
+    assert num_rows == db.table('test1').num_rows()
+
+def test_bounded_state(db):
+    pass
```
Comment (willcrichton) [inline tests/py_test.py]: We should get these tests in before merging!

File: examples/shot_detection/shot_detect.py (hunk 1 above)
Comment (apoms) [reply examples/shot_detection/shot_detect.py]: This leaves open the ability to add multiple output Ops (multiple output tables). `BulkJob(output=...)` would change to `BulkJob(outputs=[...])`

File: examples/tutorial/00_basic.py (hunk 2 above)
Comment (apoms) [reply examples/tutorial/00_basic.py]: Yup.

File: examples/tutorial/01_sampling.py (hunk 3 above)
Comment (apoms) [reply examples/tutorial/01_sampling.py]: I think we should probably have some sort of API reference... maybe something like what flask has? Specifically the API section: http://flask.pocoo.org/docs/0.12/

File: python/scannerpy/collection.py (hunk 5 above)
Comment (apoms) [reply python/scannerpy/collection.py]: They are functional, they just aren't used in specifying jobs anymore.

File: python/scannerpy/database.py (hunk 6 above)
Comment (apoms) [reply python/scannerpy/database.py]: Did they ever?

**Pull request merged**